### ISSUE TITLE
How to use result of react$? How to get props and state?

### ISSUE THREAD

**Issue by soswow** [webdriverio/webdriverio#4666]
**Describe your question with as much detail as possible**
I've gone through this blog post https://webdriver.io/blog/2019/04/03/react-selectors.html and it didn't help me. Especially the last part is not clear at all. There is a link to https://github.com/facebook/react-devtools which is deprecated and was merged into React itself. Is this a right page https://reactjs.org/docs/test-utils.html ? I've also looked through issue in this repo and could find anything related.

How do I get a react's prop from an element that is a result of `.react$` call?

**Environment (please complete the following information):**
 - **WebdriverIO version:** 5.15.1
 - **Mode:** test runner
 - **If WDIO Testrunner, running sync/async:** async
 - **Node.js version:** 10.13.0
 - **NPM version:** 6.4.1
 - **Browser name and version:** 77.0.3865.120
 - **Platform name and version:** Mac

**Comment by christian-bromann** [comment]
> How do I get a react's prop from an element that is a result of `.react$` call?

You don't. The command only allows to fetch elements based on react props and states provided by you. We can definitely update the link though.

**Comment by baruchvlz** [comment]
>  Is this a right page https://reactjs.org/docs/test-utils.html ?

No, test-utils are not the same as dev tools. React dev tools is a plugin for your browser which allows you to see the React tree in your console. We recommend using this so you can find the exact names for the components, props, state, any HOC which you might have to wrap your selector with. Again, this is recommended, not required.

__note__ There is a known bug and [open issue](https://github.com/baruchvlz/resq/issues/43) in the library WDIO uses for React selectors where using the React selectors with components wrapped around HoCs might not work properly.

The `react$/$$` commands return a WDIOElement not a React instance so you cannot get the props. With the current implementation the is no way to get the props nor state. If you need to filter your search you can do what's suggested [here](https://webdriver.io/blog/2019/04/03/react-selectors.html#selecting-and-filtering). 

With all this said, I think it's a good idea for WDIO to return a new object `WDIOReactElement` which extends `WDIOElement` and we can have then the props and state in that object.

**Comment by soswow** [comment]
@baruchvlz @christian-bromann Thank you folks. Yeah, it all makes sense.
I am playing with this function atm:
```
window.FindReact = function(dom) {
    let key = Object.keys(dom).find(key=>key.startsWith("__reactInternalInstance$"));
    let internalInstance = dom[key];
    if (internalInstance == null) return null;

    if (internalInstance.return) { // react 16+
        return internalInstance._debugOwner
            ? internalInstance._debugOwner.stateNode
            : internalInstance.return.stateNode;
    } else { // react <16
        return internalInstance._currentElement._owner._instance;
    }
}
```
With that, I can get `props` by given DOM element through hidden attributes. I am wondering how can I use this in conjunction with react$. I can execute arbitrary code via `execute` for example.
🤔

**Comment by baruchvlz** [comment]
If you need the return to contain the props and state ASAP and can't wait for the WDIOReactElement to be added, you could probably use the [library](https://github.com/baruchvlz/resq/blob/master/src/waitToLoadReact.js) used by WDIO but it doesn't support anything lower than React 16 cause it uses the FiberNodes introduced in 16.

```js
import { resq$ } from 'resq'

browser.execute(() => {
  return resq$('MyComponent', document.getElementById('root')) // root of your React application
});
```

This returns an object similar to:

```ts
interface RESQNode {
    name: 'MyComponent',
    node: HTMLElement | null,
    isFragment: boolean,
    state: string | boolean | any[] | {},
    props: {},
    children: RESQNode[]
}
```

**Comment by soswow** [comment]
I am confused a bit. This code runs on node. Callback function/argument of `execute` then stringified and sent to the browser, evaled and called over there. How should runtime of browser JS to have `resq$` in it after I imported it in node?

Using `react$` I am currently getting object like that:
```
Element {
      sessionId: '8f1352bfa57a874323329e9a94e80616',
      elementId: '0.7462215148005813-2',
      ELEMENT: '0.7462215148005813-2',
      selector: 'Editor',
      parent:
       Browser {
         sessionId: '8f1352bfa57a874323329e9a94e80616',
         capabilities:
          { acceptInsecureCerts: false,
            // .... 
            os: undefined },
         addCommand: [Function],
         overwriteCommand: [Function] },
      index: 1,
      emit: [Function: bound ],
      isReactElement: true,
      addCommand: [Function],
      overwriteCommand: [Function] }
```
Is there a way for me to go from this object to a reference to an actual DOM element inside eval'ed function using `.execute()`. Like:
```js
browser.execute(() => {
  const element = somehowGetItUsingwResultOfreact$;
  return doeSomethingWithDomElement(element);
});
```
```

**Comment by soswow** [comment]
Anyway. I think I have a way forward. The problem I have right now is actually related to the fact my component root doesn't have representation in DOM (so, only fragment is used) so I can't find it in the dom atm. But it's not related to webdriverio. Thank you.

**Comment by christian-bromann** [comment]
The problem attaching props and states to the element is that they are not updated. But I think we should provide a getter to request these values by calling execute as mentioned by @baruchvlz above.

**Comment by christian-bromann** [comment]
@soswow would the following interface work for you:

```js
const elem = browser.react$('Comp')

elem.getProp('myProp')
elem.getCurrentState()
```

We are still discussion what the best to approach the fragment issue. Any feedback is welcome!

**Comment by soswow** [comment]
@christian-bromann This would be pretty cool and useful, hands down! Is there a technical reason why the state is retrieved as a single object, but props one by one by key? Or it's not full API? Will there be smth like `.getProps()` and `.getState('myStateName')`? Also curious why state has `current` prefix, and prop doesn't?

**Comment by christian-bromann** [comment]
> Is there a technical reason why the state is retrieved as a single object, but props one by one by key?

I am not a React expert but afaik a state can technically be from any type while properties are key/value pairs.

> Also curious why state has `current` prefix, and prop doesn't?

I don't have any strong opinion on that. @baruchvlz what do you think? I could live with `getProps` and `getState` too.

**Comment by soswow** [comment]
> a state can technically be from any type while properties are key/value pairs.
I've been working exclusively with TypeScript for the past couple of years, so for us React's Component has strongly typed State and Props. Sure, for vanila JS there is `prop-types` library with pretty limited abilities and no such thing exist for the state. You might be right here.

**Comment by christian-bromann** [comment]
> think a simpler solution would be to have two methods with no arguments `getProps` and `getState` which would return their respective values from the component since I don't really see the use of getting one specific prop (or state value) off a component.

This sounds reasonable to me.

**Comment by shaharyarmaroof-polarr** [comment]
Is there any update on this? It seems that this is still an open issue.

**Comment by christian-bromann** [comment]
@shaharyarmaroof-polarr an open issue with the label `help wanted` so if you are interested in this please get involved. Have a read on our [contribution guidelines](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) and let us know if you have any questions.

**Comment by sanyco92** [comment]
Well, almost 2 years have passed and no one has implemented this feature?

**Comment by srinivasanke** [comment]
Hello All,

IS there any way pass props values to react customer element (Parent to child)
